src/readii/io/writers/base_writer.py

```diff
@@ -0,0 +1,171 @@
+import re
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from datetime import datetime, timezone
+from pathlib import Path
+from types import TracebackType
+from typing import Any, ClassVar, Dict, Optional, Tuple
+
+from imgtools.dicom.sort.exceptions import InvalidPatternError
+from imgtools.dicom.sort.parser import PatternParser
+
+from readii.utils import logger
+
+
+@dataclass
+class PatternResolver:
+	"""Handles parsing and validating filename patterns."""
+
+	filename_format: str = field(init=True)
+
+	DEFAULT_PATTERN: ClassVar[re.Pattern] = re.compile(r"%(\w+)|\{(\w+)\}")
```

Comment:
Can you add a comment breaking down what this regex is at a high level?

Reply:
yup, moved to a utils directory and added more documentation throughout the module

Let me know if this is not confusing:

https://github.com/bhklab/readii/blob/cdfcdfcf1ef54cf3aeaa20d94cfaf3610f49792d/src/readii/io/utils/pattern_resolver.py